src/Raven.Server/Documents/RevisionsBinCleaner.cs

```diff
@@ -0,0 +1,172 @@
+{
+    public class RevisionsBinCleaner : BackgroundWorkBase
+    {
+        private readonly DocumentDatabase _documentDatabase;
+        private readonly RevisionsBinConfiguration _configuration;
+        private readonly long _batchSize;
+
+        public RevisionsBinCleaner(DocumentDatabase documentDatabase, RevisionsBinConfiguration configuration) : base(documentDatabase.Name, documentDatabase.DatabaseShutdown)
+        {
+            _documentDatabase = documentDatabase;
+            _configuration = configuration;
+            _batchSize = _documentDatabase.Is32Bits ? 1024 : 10 * 1024;
+        }
+
+        protected override async Task DoWork()
+        {
+            await WaitOrThrowOperationCanceled(_configuration.RefreshFrequency);
+
+            await ExecuteCleanup(_configuration);
+        }
+
+        public static RevisionsBinCleaner LoadConfigurations(DocumentDatabase database, DatabaseRecord record, RevisionsBinCleaner oldCleaner, string nodeTag)
+        {
+            try
+            {
+                var config = record.RevisionsBin;
+
+                if (config == null || config.Disabled)
+                {
+                    oldCleaner?.Dispose();
+                    return null;
+                }
+
+                if (oldCleaner != null && config.Equals(oldCleaner._configuration))
+                    return oldCleaner;
+
+                oldCleaner?.Dispose();
+
+                var cleaner = new RevisionsBinCleaner(database, config);
+                cleaner.Start();
+
+                if (cleaner.Logger.IsInfoEnabled)
+                    cleaner.Logger.Info($"Executed revisions-bin cleanup on {database.Name} started");
+
+                return cleaner;
+            }
+            catch (Exception e)
+            {
+                const string msg = "Cannot enable revisions-bin cleaner as the configuration record is not valid.";
+                database.NotificationCenter.Add(AlertRaised.Create(
+                    database.Name,
+                    $"Revisions-bin cleaner error in {database.Name}", msg,
+                    AlertType.RevisionsConfigurationNotValid, NotificationSeverity.Error, database.Name));
+
+                var logger = LoggingSource.Instance.GetLogger<RevisionsBinCleaner>(database.Name);
+                if (logger.IsOperationsEnabled)
+                    logger.Operations(msg, e);
+
+                return null;
+            }
+        }
+
+        internal bool IsFirst => AbstractBackgroundWorkStorage.ShouldHandleWorkOnCurrentNode(_documentDatabase.ReadDatabaseRecord().Topology, _documentDatabase.ServerStore.NodeTag);
+
+        internal async Task<long> ExecuteCleanup(RevisionsBinConfiguration config = null)
+        {
+            var totalDeletedEntries = 0L;
+
+            config ??= _configuration;
+
+            if (config == null ||
+                config.MinimumEntriesAgeToKeep == null ||
+                config.MinimumEntriesAgeToKeep.Value == TimeSpan.MaxValue ||
```

Comment:
what happened if `MinimumEntriesAgeToKeep` is a negative value?

Reply:
Treated as TimeSpan.Zero (deletes all)